README
QUILT

Written by Philip Lijnzaad (while at EBI; now at the Princess Maxima Centrum for
Pediatric Oncology (<p.lijnzaad@prinsesmaximacentrum.nl> or <plijnzaad@gmail.com>)

Quilt is a program to delineate hydrophobic patches on protein surfaces. For the
description of the algorithm, please quote/refer to

  Lijnzaad, P., Berendsen, H.J.C. and Argos, P. (1996) 
  "A method for detecting hydrophobic patches on the protein surfaces."
  Proteins: Structure, Function, Genetics 26:192-203

For a statistical overview of patches on monomeric proteins, see:

  Lijnzaad, P., Berendsen, H.J.C. and Argos, P. (1996) 
  "Hydrophobic patches on the surfaces of protein structures." 
  Proteins: Structure, Function, Genetics, 25:389-397

For a statistical overview of patches on the interface of obligate oligomeric
proteins, see:

  Lijnzaad, P. and Argos, P. (1997)
  "Hydrophobic patches on protein subunit interfaces: characteristics and prediction."
  Proteins: Structure, Function, Genetics, 28:333-343
  (doi:10.1002/(SICI)1097-0134(199707)28:3<333::AID-PROT4>3.0.CO;2-D)


To apply Quilt to a complete molecular dynamics trajectory of a protein (such
as e.g. produced by GROMACS), be sure to check out the package PatchTrack, as well
as the following paper:

  Lijnzaad, P., Feenstra, K.A., Heringa, J. and Holstege, F.C.P. (2008)
  "On Defining the Dynamics of Hydrophobic Patches on Protein Surfaces"
  Proteins, Structure, Function, Bioinformatics 72: 105-114.


I recently noticed that there now is completely different tool called quilt.
It is software for managing software patches (incremental source code
differences). It has nothing to do with protein structure analysis.

Installing:
----------

See the INSTALL file.  I recently put this thing on github, so
I simplified things a bit. Please fork the project and send me pulll
requests!

PDB files are searched for in the path given enviroment variable PDBPATH,
which can be a colon separated list.

Documentation:
-------------

Currently this README file is the main documentation; along with the
publication it should be enough to get started. Also, be sure to read the
output of `quilt -h'

A list of files contained in this distribution, along with a brief
description, is given in file FILES; this may help a bit as well.

Please note that the package contains a number of un(der)documented features.
Especically the full triangulation routines are not usable in the current
version. 

The directory examples contains the example in- and output of running the
program on lysozym; the shell script in scripts/job is how it was
created. The scripts (and most other stuff) often have a -h option, and/or
are commented fairly extensively.

Parameters:
-----------

See quilt -h and the article. Also see .residues and .atoms for atom
radii. One thing to play with is the polar expansion radius. A reasonable
setting is in the 1.0 - 1.4 range. The larger this value, the smaller but
more prominent the patches will be; smaller values give more disperse
patches, that are larger and may contain the occasional polar atom.

Input:
------
A pdb file, which is searched for in the environment variable PDBPATH. In it,
it searches for 3fxn.brk or 3fxn.brk (alter pdb.h if necessary). The program
can also read a file in the format N\nN * {X Y Z RADIUS\n}, and can also read
DSSP files (see also pdb.h and pdb.c; alter DSSPPATH etc. if necessary).

Output: 
------- 

progress/warnings/errors to STDERR, patches to STDOUT.  The stderr output can
largely be ignored; most warnings are informational or inconsequential.  The
patches, written to STDOUT, are sorted by descending patch size. If
recovering is switched on (-R option), there is first a report on the
correspondence between the rankings  looking like: 

  # new patch ranks, after recovering
  #0: 68 (850.487) WAS #0: 32 (289.169)
  #1: 80 (815.522) WAS #1: 39 (264.779)
  #2: 56 (770.897) WAS #4: 25 (238.075)

is given; you can ignore this. After this, a list of patches is given. Each
patch looks like:

# 2 14 atoms 16 patches 365 points area 198.620 0 bndries ABC 11.638 5.688 1.847
% 599 616 618 831 835 840 844 845 854 855 858 1065 1069 1070 
 G 79 @ CA =3198-;K 81 @ CG =586-;K 81 @ CE =4088-;G 107 @ C =586-;C 108 @ C =111-;
 V 109 @ CA =334-;V 109 @ CG1=3915-;V 109 @ CG2=3269-;V 111 @ CA =389-;V 111 @ C =151-;
 V 111 @ CG1=825-;I 138 @ CA =218-;I 138 @ CG1=2067-;I 138 @ CG2=935-;

All this means the following: patch nr 2 (i.e. 3rd in rank; they start at 0)
consists of 16 atomic faces (the faces used to be called patches, I know this
is confusing!)  on 14 different atoms, totalling 365 sampling points and
having a total area of 198.620 square Angstroms of solvent accessible
surface. There are 0 boundaries (as this is not implemented) and the
semi-axes A,B and C of an 3D ellipsoid fitted to the patch are given in
Angstroms. Then (the lines starting with '% ') follow the atom numbers (the
order in the file; counting starts at 0), followed by the textual atom IDs
which are separated by semi-colons. The first atom of this patch is CA of
Gly79 of chain ' ', having an individual solvent accesible surface area of
31.98 square angstroms (numbers are multiplied by 100), and no ('-')
secondary structure information is available (the program can integrate data
from the DSSP secondary structure assignments, but that's not done for the
test run).

A more formal description would be:

'#' RANK NATOMS NFACES NPOINTS 'area' AREA '0 bndries' 'ABC' A B C 
'%' ATOMNO
A_ACID CHAIN RESIDUENO INSERTION_CODE '@' ATOMNAME '=' AREA*100 SEC.STR. ';'


Example:
--------
A typical way of running things (Bourne shell; see also scripts/job) is:

  quilt  -n 252 -ep 1.4 -R -p 8lyz.pdb -a 8lyz.area > 8lyz.pat 2> 8lyz.log

252 sampling points per atom are used, a polar expansion radius of 1.4
Angstrom is used, and the preliminary patches are recovered after expansion.
The -a option writes a PDB file `8lyz.area' which has the atomic solvent
accessabilty in the fifth field. This is needed for the following.

Randomization:
--------------

To get an idea of the significance of the patches, you need to randomize the
surface, and compare the distribution of patch sizes with the actual one.
First, the actual distribution of patch sizes is histogrammed by extracting
the patch sizes and running them through the perl script `histo' which can be
found in the scripts/ directory:

  awk '/^# [0-9]/{print $10}' 8lyz.pat | histo -l 0 -s 10 > 8lyz.histo

(bins starting at 0, and bin size 10 Angstroms; see histo -h for more
documentation)

Next, you randomize the structure several times. For this you need a pdb file
with atomic solvent accessabilities such as produced with the -a option
(failing to do this will give a wrong hydrophobic surface fraction, resulting
in a bias that makes large real patches look small compared to the randomized
version).  On this file, you run the program N times (say between 20 and 100),
with the -ran option. Output is caught in different files, and an `average
distribution' is made simply by doing a histogram of sizes on all the
randomized files at once:

  for i in `seq 1 100`
  do 
    quilt -ran -n 252 -ep 1.4 -R -p ./8lyz.area > 8lyz.ran$i 2> /dev/null
  done
  
  cat 8lyz.ran[0-9] | awk '/^# [0-9]/{print $10}' | histo -l 0 -s 10 \
    | awk '{print $1, $2/50}' > 8lyz.ranhisto

The results are best viewed as two graphs in the same plot, using a graphical
plot program, e.g. using xmgrace -nxy 8lyz.histo -nxy 8lyz.ranhisto. While
not statistically rigourous, it does show which patches appear to be
unusually large for the given structure.

This whole process, including the randomization, is automated and documented
in scripts/one-structure.sh. 


Results:
--------
The perl script `patresidues' is provided to convert the output patches to
something slightly more readable. Run as

  patresidues < 8lyz.pat > 8lyz.txt

The script Rpatsel produces a little rasmol script for easier viewing; run 
as 

  for i in 0 1 2 3 4 
  do 
    Rpatsel 8lyz.pat $i  
  done > 8lyz.rml

Loading the script 8lyz.rml into Rasmol will define p0 to be patch number 0,
p1 is patch no 1, etc. You can then 'select p2' and color or display it the
way you want. If there are cofactors present, you should rename them as they
have been assigned funny oneletter-codes by the ~/.residue template file;
Rpatsel has a provision for doing this (see Rpatsel -h)

That's about it. I'd be very pleased to hear about any experience, good
or bad, with the programs and scripts that make up the  package. Questions,
comments, bug reports please to plijnzaad@gmail.com. Alternatively, just
fork the project and send me pull requests!


								Philip


ChangeLog
2021-11-03  Philip Lijnzaad  <plijnzaad@gmail.com>

	* v1.3: Got rid of the template spheres caching as it causes problems on Mac.
	(thanks, https://github.com/jacakrj1 !!) Number of points is now
	default 252 (which is the highest, no need to pass it.

2020-09-24  Philip Lijnzaad  <plijnzaad@gmail.com>

	* v1.2: revamped the release process: no more releases, just git.
	Segfaults on Mac OSX though.

2014-11-13  Philip Lijnzaad  <philip@mayr.op.umcutrecht.nl>

	* Improvements to the release script, including make'ing it. 
	Version bump to version 1.1h

2014-08-18  philip  <philip@laptop-philip>

	* got rid of old getopts.pl.

2014-01-20  philip  <philip@philip-laptop>

	* scripts/patresidues: made to work with perl 5.9 regexps

	* added initializer on utils/utils.c:__file__ and __line

2011-11-15  philip  <philip@philip-laptop>

	* Added the full citation details for the PatchTrack paper
	referred to in the README.

2009-11-30  Philip Lijnzaad  <p.lijnzaad@umcutrecht.nl>

	* 1.c: ulong problem. Changed e-mail adresses. Now using 50
	randomizations. Added defs for ulong and FILENAME_MAX for
	compilation under MacOSX (unclear if works)

2007-02-27  admin  <p.lijnzaad@med.uu.nl>

	* 1.1b: clean up of the build-distrib.sh system and some general
	        polishing, no big changes.


src/quilt.h
/* ------------------------------------------------------------------ *
 *								      *
 * Permission to copy all or part of this work is granted, provided   *
 *  that this NO WARRANTY and this copyright notice are retained      *
 *  verbatim and are displayed conspicuously.  If anyone needs other  *
 *  permissions that aren't covered by the above, please contact the  *
 *  author.							      *
 *								      *
 * NO WARRANTY: THIS WORK IS PROVIDED ON AN "AS IS" BASIS.  THE       *
 *  AUTHOR PROVIDES NO WARRANTY WHATSOEVER, EITHER EXPRESS OR         *
 *  IMPLIED, REGARDING THE WORK, INCLUDING WARRANTIES WITH THE WORK,  *
 *  INCLUDING WARRANTIES WITH RESPECT TO ITS MERCHANTABILITY OR       *
 *  FITNESS FOR ANY PARTICULAR PURPOSE.				      *
 *								      *
 * Philip Lijnzaad, plijnzaad@gmail.com
 *								      *
 * ------------------------------------------------------------------ */

#ifndef _QUILT_H_
#define _QUILT_H_ 1

#include "compat.h"

#include "utils.h"			/* for uchar etc. typedefs */
#include "list.h"			/* for list typedefs */
#include "pdb.h"			/* for pdb typedefs */

#ifdef __GNUC__				/* gnu cc, knows directive 'inline' */
#  ifdef DEBUG
#  define inline
#  endif
#else  					/* others don't */
#  define inline
#endif

/* constants: */
/* floats (tolerances etc.) */
#define DEF_PROBE_RADIUS 1.40		/* default radius of probe */

#define COS_SMALL_ANGLE 0.99		/* cosine of 8.1 degrees. Used in */
					/* zip_pair to rule out too sharp */
					/* corners */ 
/* tolerances: */
#define TINY_ANGLE_TOL 0.005		/* fractional tolerance for */
					/* judgeing an angle too tiny */
#define INTERSECT_TOL 0.001		/* atom overlap should be at least */
					/* this for intersect_pair() to be */
					/* called */

#define SWALLOW_TOL 0.001		/* if the difference with being */
					/* completely swallowed is less than */
					/* this, consider it to be so */

#ifdef SPH3
#  define MAX_NGBS_PER_PNT 7		/* maximal coordination number */
#else
#  define MAX_NGBS_PER_PNT 6		/* maximal coordination number */
#endif /* SPH3 */

#define NDAT_ITEMS 3			/* 3 coordinates per point */

/* types of spheres: */
#define SPH_ICOS    0			/* icosahedron */
#define SPH_PDOD    1			/* pentakisdodecahedron */
#define SPH_ZIEL    2			/* for sph3. (not used actually) */
#define SPH_UNKNOWN 3			/* when read from file */

#define N_TMPLT_SPHERES 7		/* number of template spheres */
#define MAX_TMPLT_PNTS 256 /* 252 */	/* highest number of points on */
					/* template spheres (== 10*d*d+2; */
					/* 30*d*d+2 if dodecaedron-based) */
#define MAX_PATHLENGTH MAX_TMPLT_PNTS/2
#define MAX_NCELLS 343			/* max nr of cells per sphere */
#define MINRADIUS 3*2			/* index in limits array */
#define MAXRADIUS 3*2 + 1		/* index in limits array */

#define MAX_NNGBS 256			/* max nr of neibors to any atom */

#define MAX_NCONS 16			/* max nr. of cons out of a point */

/* flags  ( #define BIT(n) (1<<(n)) is already defined) */
/* general purpose flag */

/* flags for points */
#define BURIED 		BIT(0)		/* points/edges buried */
					/* DON'T CHANGE THIS ONE !!! */
					/* set in acc.c:calc_acc() */
#define BNDRY		BIT(1)		/* part of a bndry */
					/* set in top.c:find_bndry(), */
					/* top.c:hide_flags()  */
#define PATCH		BIT(2)		/* marked as part of patch */
					/* set in top.c:find_bndry(), */
					/* top.c:mark_interior()  */
#define ISOLATED	BIT(3)		/* single pt bndry & patch */
					/* for shell: completely exposed ! */
					/* set in top.c:find_bndry() */
#define TWOWAY		BIT(4)		/* bndry passes twice (or more) */
					/* through through this
                                         * pt. Happens if the boundary is
                                         * around a small exposed 'strip' of
                                         * surface, that happens to be
                                         * sampled by just one point 
                                         */
#define UTURN		BIT(5)
#define SHARED		BIT(6)		/* point shared by 2
                                         * bndries. Sort of opposite of
                                         * TWOWAY: happens where there is
                                         * a patch with two (or more)
                                         * bndries, where there is small
                                         * strip sampled by one point
                                         * which is used by both bndries
                                         * (in opposite directions) 
                                         */
#define UNWANTED 	BIT(7)		/* to be neglected */

/* flags for bndries, patches, and/or shells */
#define COORDINATES	BIT(8)

/* #define BURIED	BIT(0)  should be same everywhere */
#define DONE		BIT(0)		/* patch has been traversed once */

#define ISWIDOW		BIT(1)		/* no neighbours (shell only) */
					/* set in acc.c:calc_acc() */
#define HASWIDOWS	BIT(2)		/* atom contains a widow point */
/* #define ISOLATED	BIT(3)		 completely exposed atom */
					/* set in acc.c:purge_ngbs(), */
					/* acc.c:isolated_atom()  */
#define MBNDRIES	BIT(4)		/* shell contains > 1 bndrys */
					/* top.c:find_patches */
#define MPATCHES	BIT(5)		/* shell contains > 1 patches */
#define COMPLEX		(MBNDRIES | MPATCHES) /* both */
#define POLAR		BIT(6)		/* atom is polar */
#define REMOVE_WIDOWS	BIT(8)		/* to flag need for remove_widows() */

/*  */
#define RETOP 		BIT(9)		/* shell has to triangulated again */
#define STITCHED	BIT(10)		/* shell does not have to be "  " */

#define INVALID_SHELL(S)  ( ((S)->flags) & (BURIED | ISOLATED | ISWIDOW) )
#define VALID_SHELL(S)  (! INVALID_SHELL((S)))

/* macros: */
#define FASTMOD_NTIMES 3		/* nr of times to repeat row of 0..n */

#define FASTMOD(W, NGBS) fast_mod[ (NGBS) ][ (W) ] 

/* #define MOD(A,B) FASTMOD((A),(B)) */

#define MOD(A,B) ((A)%(B))		/* fast_mod thing is bugged */

/* int fast_mod_func(int, int); */
/* #define MOD(A,B) fast_mod_func((A),(B)) */

/* typedefs: */
#include <limits.h>

typedef unsigned short SHORT;
#define POINT_MAX USHRT_MAX

/* to obtain +- 25 % reduction in memory usage, use following instead: */

/* typedef unsigned char SHORT;
 * #define POINT_MAX UCHAR_MAX
 */

typedef unsigned short EDGE_ID;
#define EDGE_MAX USHRT_MAX

#define AVAIL (POINT_MAX)		/* can't use that as point id */
#define UNAVAIL (POINT_MAX-1)		/* neither this */
#define MAXNROFPOINTS MIN((POINT_MAX-2), ((EDGE_MAX+6)/3))
					/* since 3N - 6 edges !! */
typedef struct point_ {
  float *xyz;
  float *ngbs_xyz[MAX_NGBS_PER_PNT];	/* coords of ngb points */
  SHORT nr;				/* its number; mainly for debugging */
  SHORT flags;
  SHORT nngbs;				/* number of neibors or edges */
  SHORT ngbs	 [MAX_NGBS_PER_PNT];	/* ids of ngbs */
  EDGE_ID edges	 [MAX_NGBS_PER_PNT];	/* ids of edges */
/*  was: SHORT edges [MAX_NGBS_PER_PNT]; */
  SHORT whoamis  [MAX_NGBS_PER_PNT];	/* index of this point in its */
  float area;
  struct point_ 
    *Ngbs	 [MAX_NGBS_PER_PNT];	/* pointers to neighbour points */
} point_t, *point_p;

typedef struct edge_ {			/* little used, but available */
  SHORT ends[2];			/* endpoints; always ends[0]<ends[1] */
  SHORT flags; 
  union sw_ {
    SHORT s; 
    uchar w[2]; 
  } sw; 
#define face_with sw.s			/* when doing find_faces */
#define whoami sw.w			/* othewise */

  float direction[3];
#define midpoint direction		/* needed during tr._sphere() */
  float length2;
  float arclen;
} edge_t, *edge_p;

typedef struct face_ {			/* not used, but available */
  ushort edges[3];
  ushort points[3];
/* was:  SHORT edges[3]; */
  SHORT flags;
  float area, midpoint[3];
} face_t, *face_p;

typedef struct sphere_ {
  int npoints, nedges, nfaces, order, maxringlen;
  int dim;				/* dimension of grid */
/*  int ncells; */
  int type;				/* SPH_ICOS or SPH_PDOD */

/*   point_p * cells;			/# array of length dim^3, holding #/
 * 					/# linked lists of poinst in cell #/
 *   uchar * pts_per_cell;			/# nr of atoms per cells #/
 */

  float * coordinates;			/* coordinates of all points, as  */
  point_p Points;			/* series of float[3]'s */
  edge_p Edges;
  face_p Faces;

  double edgelengths[3];		/* min, max, ideal edge length */
  double arclengths[3];			/* min, max arclen, ideal length */
  double beta, halfalpha;		/* cos(alpha)/cos(alpa/2); alpha/2 */
  double sinbeta;
  /* halfalpha = 0.5*maxarclen, beta is cos(minarclen)/cos(halfalpha) */
#define ideal_arclen arclengths[2]

} sphere_t, *sphere_p;

typedef struct bndry_ {			/* boundary of a patch */
  struct bndry_ *next;			/* next bndry belonging to patch */
  struct patch_ *patch;			/* patch it belongs to */
  SHORT *path;				/* point numbers */
  list_p * cons;			/*  */
  union { 
    struct shell_ ***ngbs;		/* during connect_2_bndries */
    float * normals;			/* during stitch */
  } pt;					/* shells (NULL terminated) */
  union { 
    struct shell_ ** shell;		/* during connect_bndries() */
    float *coordinates;			/* during stitch() */
    struct bndry_ ** bndry;		/* during connect_elements() */
  } ngbs;

  SHORT length;				/* length of path */
  SHORT ncons;				/* nr of connections */
  SHORT nngbs;				/* nr of bndry ngbs */ 
  SHORT flags;
} bndry_t, *bndry_p;

typedef struct patch_ {			/* only on one atom */
  struct patch_ * next;
  bndry_p bndries;
  atom_p atom;				/* to which it belongs */
  struct patch_ ** ngbs;		/* array of ngbs  */
  SHORT nngbs;				/* nr of ngbs */
  SHORT npoints;			/* all points of this patch */
  SHORT internals;			/* just the internals */
  SHORT flags;
  SHORT * points;			/* only internal points ! */
  float area;				/* area as if R where 1.0 ! */
} patch_t, *patch_p;

typedef struct shell_ {                 /* partial triangulation of atom;
                                           holds all the additional
                                           information per atom */ 
  union {
    struct shell_ * next;		/* next atom in grid cell */
    list_p *** conlists;
  } next;
  sphere_p sphere;			/* what sort of sphere */
  atom_p atom;				/* the atom it belongs to */
  struct shell_ ** ngbs;		/* neigbouring shells */
  patch_p patches;
  SHORT *pointflags;			/* point->flags per point */
  SHORT nngbs;				/* nr of neibors */
  SHORT flags;				/* eg. BURIED */
  SHORT nburied;			/* nr of atoms buried */
  float area;				/* area if it were a unit sphere */
} shell_t, *shell_p;

typedef struct surf_ {			/* usually covers more than one atom */
  struct surf_ * next;
  int npatches;				/* nr of patches forming surface */
  int npoints;				/* nr of points sampling it */
  patch_p *patches;			/* array of ptrs to patches */
  atom_p *atoms;			/* array of ptrs to atoms (unique) */
  double area;				/* calc'd in sum_area; real radii now*/
  int natoms;				/* nr. of unique atoms */
  int rank;				/* rank in sorting */
  struct surf_ * old;			/* pointer to equivalent surf before */
  float ellips_axes[3];
  int nbndries;
  list_p bndries;
} surf_t, *surf_p;			/* rescovering buried area */

typedef struct box_ {
  atom_p atoms;
  int natoms;
  structure_p structure;
  float * original_radii;
  shell_p * cells;
  shell_p shells;
  shell_p old_shells;			/* for 2nd time around ... */
  int ncells;
  int dims[3];				/* dimensions: nr of x,y and z boxes */
  float maxrad;				/* maximum radius */
  float spacing;
  float limits[8];			/* min-x, max-x; min-y, max-y; etc. */
  int nburied;
  double totarea,Narea,Oarea,Carea,Sarea;
  char name[20];
} box_t, *box_p;

#ifdef undefined
typedef struct con_ {
  bndry_p bndries[2];
  SHORT offset[2];
  float direction[4];			/* x, y, z (normalized), r */
} con_t, *con_p;
#endif /* old version */

typedef struct con_ {
  bndry_p bndries[2];
  struct {				/* rever to as con->flags[1].offset */
    SHORT seen:1, done:1, offset:14;
  } flags[2];
  float direction[4];			/* x, y, z (normalized), r */
} con_t, *con_p;

typedef struct seam_ {			/* old type; had to keep it small */
  con_p con;
  short end, length; 
  short from, till;
  bndry_p bndry;			/* always bndry==con->bndries[end] */
  SHORT *first, *last;
} seam_t, *seam_p;

typedef struct br_ {
  struct br_ * next, *prev;
  list_p *where;
  bndry_p bndry;
  shell_p shell;
  int offset;
  const float * coord, *normal;
  uint flags;
  list_p *conlist;
  list_p **conlists;
  ushort *pointflags;
  double atomradius;
  double dir[4]				/* unit vector along established bit */
    ,newdir[4], 			/* unit vector along candidate con */
    phi,				/* angle the seam makes here */
    mindotp, dotp[2];			/*  */
} br_t, *br_p;

/*** global numbers, data ***/
/* in main.args: (options) */
extern int opt_careful, opt_maximize, opt_warn, opt_connect_touch, debug,
	opt_general;
/* extern atom_p root_of_atoms;		/# for calculating index #/
 * extern shell_p root_of_shells;
 */

extern FILE * connections_file;

typedef struct radius_param_ {
  double 
/*    radius, OBSOLETE */
    extension,
    reduction,
    polar_extension,
    polar_reduction;
} radius_param_t;

extern radius_param_t radius_param;

/* #define opt_radius          (radius_param.radius) OBSOLETE */
#define opt_extension 	    (radius_param.extension)
#define opt_reduction 	    (radius_param.reduction)
#define opt_polar_extension (radius_param.polar_extension)
#define opt_polar_reduction (radius_param.polar_reduction)

/* in misc.c: */
extern FILE  *logfile;

extern int overall_npatches;
extern int maxnpoints;			/* set in choose_(uni)_sphere */
extern int ** fast_mod;			/* matrix for speeding up modulo-op. */

/* in templates.c: */
extern sphere_p Template_spheres[ N_TMPLT_SPHERES ];
extern int Template_npoints[ N_TMPLT_SPHERES ];

#define TEMPLATE_NPOINTS_CONTENTS  { 12, 32, 42, 92, 122, 162, 252  }
/* this is mixture of Sloan's A005901 and another one for the pdod (which I
   can't find */

#define TEMPLATE_NPOINTS_HALFWAY     { 22, 37, 64, 102, 142, 204 }

extern int Template_based_on[ N_TMPLT_SPHERES ];
#define TEMPLATE_BASED_ON_CONTENTS { \
  SPH_ICOS, SPH_PDOD, SPH_ICOS, SPH_ICOS, SPH_PDOD, SPH_ICOS, SPH_ICOS  }

extern int Template_order[ N_TMPLT_SPHERES ];
#define TEMPLATE_ORDER_CONTENTS  { 1,  1,  2,  3,   2,   4,   5  }

/*** prototypes: ***/
/* in io.c: */
box_p get_atoms(FILE * pointsfile, char* protein, char*binspec, uint pdbflags);
void get_template_spheres(void);

void print_atom(FILE*, atom_p, char*atomcolor, char*conn_color);
void write_atom(char*filename, atom_p, char*atomcolor, char*conn_color);
void print_cons(FILE*, atom_p, char *color);
void write_cons(char*filename, atom_p, char *color);
void print_sphere(FILE*, atom_p, char *color);
void write_sphere(char *filename, atom_p, char *color);
void print_triangulines(FILE*, box_p);

int print_atom_dots(FILE*, atom_p, char*prepend, char*append);
int print_dots(FILE*, box_p box);
void print_areas(FILE*, box_p box);

/* in templates.c: */
void generate_template_spheres(void);
void randomize_points(sphere_p Sphere);

sphere_p tesselate(int type, int order);
void triangulate_sphere(int opt_c, sphere_p);

sphere_p recursion_tessel(int type, int freq);
void free_templates(void);

/* in acc.c: */
int accessability(box_p box);
void free_accessability(box_p box);


/* in top.c: */
int atom_topology(box_p, int save_internal_points);
void clear_top_flags(shell_p);
int find_patches(shell_p, int opt_recover, int *nbndriesp, int firsttime);
int delete_bndry(bndry_p);
int remove_widows(shell_p shell, int *ndeletedp);
int free_bndry(bndry_p);
int free_patch(patch_p patch);
int free_patches(patch_p patches);
void free_topology(box_p);
void free_neibordata(shell_p);
void free_all_neibordata(box_p);

/* in conn.c: */
con_p alloc_con(const con_t *oldcon);
void free_con(con_p);

/* con_p save_con(bndry_p a, int offsa, bndry_p b, int offsb,
 * 	       double *dir, double d2, double rb);
 */
/* int remove_bndry_con(bndry_p); */
/* int delete_conref(con_p con, int end); *//* one side; just the refenrence*/

int remove_con(con_p con);		/* both */
int remove_con_halfknown(list_p, bndry_p); /* if one side known  */

list_p * con_place(con_p toadd, bndry_p bndry, int offset);
int connections(box_p box, int mode);
int widow_check(shell_p, int mode);
void free_connections(box_p box);

/* in seam.c: */
const float * bndry_coordinates(bndry_p bndry);
int find_seam(con_p firstcon, int firstend, int * nseamsp, 
	      int maxnseams, seam_p seams);
/* void set_seams(int nseams, seam_p seams); */

/* int do_pair(seam_p seams, int sort); */
/* int do_simple(shell_p shell); */

/* in stitch.c: */
int stitch(box_p);

/* in misc.c: */
/* int ** fast_modulo_matrix(int from, int to); */

sphere_p choose_uni_sphere(int pts_per_atom);
sphere_p choose_sphere(atom_p atomp, double density);
const char * atom_spec(box_p box, int i);

/* in surf.c: */
int print_faces(FILE *file, box_p box);
int print_surfaces(FILE*, surf_p Surfaces, const char *nature);
surf_p get_surfaces(box_p box, const char * nature);
  /* returns a list of 'surfaces', each of which is a contiguous piece of */
  /* surface of specified nature. For easy access, a surf */
int surftop(surf_p surf);
void free_surface(surf_p);
int recover(FILE*, box_p, int patdots);

/* in randomize.c: */
int ran_atoms(box_p box, int percentage);

#endif /* _QUILT_H_ */
